ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,417 @@
+                _flags.ss_glide_initial = 0;
+            }
+
+            //run airspeed/attitude controller
+            helispdhgtctrl->set_dt(G_Dt);
+            helispdhgtctrl->update_speed_controller();
+
+            //retrieve pitch target from helispdhgtctrl 
+            _pitch_target = helispdhgtctrl-> get_pitch();
+
+            //update controllers
+            _flags.bad_rpm = arot->update_hs_glide_controller(G_Dt); //run head speed/ collective controller
+            //attitude controller is updated in navigation switch-case statements
+
+            break;
+        }
+
+        case FLARE:
+        {
+            break;
+        }
+
+        case TOUCH_DOWN:
+        {
+            if (_flags.touch_down_initial == 1) {
+                //Functions and settings to be done once are done here.
+
+                #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+                    gcs().send_text(MAV_SEVERITY_INFO, "Touch Down Initiated");
+                #endif
+
+                //Set limits of z velocity and z acceleration
+                pos_control->set_max_speed_z(-800, 500);
+                pos_control->set_max_accel_z(200000);
+
+                //Get height that touchdown phase starts is initiated at
+                _z_touch_down_start = curr_alt;
+
+                //Set time that touchdown phase is initiated at
+                _t_touch_down_initiate = now;
+
+                //Get collective aggression from parameter
+                _collective_aggression = 0.15f;
+
+                _flags.touch_down_initial = 0;
+            }
+
+            //Calculate desired z
+            _des_z = _z_touch_down_start * 0.1f * expf(-_collective_aggression * (now - _t_touch_down_initiate)/1000.0f); // + terrain_offset (cm)
+
+            pos_control->set_alt_target(_des_z);
+
+            pos_control->update_z_controller();
+
+            break;
+        }
+
+        case BAIL_OUT:
+        {
+        if (_flags.bail_out_initial == 1) {
+                //functions and settings to be done once are done here.
+
+                #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+                    gcs().send_text(MAV_SEVERITY_INFO, "Bailing Out of Autorotation");
+                #endif
+
+                //Set bail out timer remaining equal to the paramter value
+                if (_param_bail_time < BAILOUT_RAMP_TIME) {
+                    bail_time_remain = BAILOUT_RAMP_TIME;
+                } else {
+                    bail_time_remain = _param_bail_time;
+                }
+
+                //set initial target vertical speed
+                _desired_v_z = curr_vel_z;
+
+                //Initialise position and desired velocity
+                if (!pos_control->is_active_z()) {
+                    pos_control->relax_alt_hold_controllers(arot->get_last_collective());
+                }
+
+                //Get pilot parameter limits
+                float pilot_spd_dn = -get_pilot_speed_dn();
```

Comment:
could put a "const" in front.. totally not necessary though, some devs think adding the const is unnecessary and a waste of time.

Reply:
done as suggested